### ISSUE TITLE
Asciidoc/Asciidoctor support

### ISSUE THREAD

**Issue by azu** [textlint/textlint#116]
We want to support [AsciiDoc](http://www.methods.co.nz/asciidoc/ "AsciiDoc")/[Asciidoctor](http://asciidoctor.org/ "Asciidoctor") format as a plugin, but there are not parser.

## Reference

- [Asciidoctor :: Discussion - Asciidoc syntax definition](http://discuss.asciidoctor.org/Asciidoc-syntax-definition-td1920.html)
- [Issues · asciidoctor/asciidoc-grammar-prototype](https://github.com/asciidoctor/asciidoc-grammar-prototype/issues)
- [Asciidoctor :: Discussion - Asciidoc syntax definition](http://discuss.asciidoctor.org/Asciidoc-syntax-definition-td1920.html)
- [Issues · asciidoctor/asciidoc-grammar-prototype](https://github.com/asciidoctor/asciidoc-grammar-prototype/issues)

**Comment by azu** [comment]
WIP https://github.com/azu/textlint-plugin-asciidoc

- [ ] Covert textmate bundle to tokens
- [ ] Create AST from tokens. it's difficult! :(

**Comment by azu** [comment]
Experimental release https://github.com/azu/textlint-plugin-asciidoc-loose

**Comment by seikichi** [comment]
I created a new textlint asciidoc plugin with [asciidoctor.js](https://github.com/asciidoctor/asciidoctor.js) AST ([seikichi/textlint-plugin-asciidoctor](https://www.github.com/seikichi/textlint-plugin-asciidoctor)).

Although there are already some plugins, there are not maintained... (textlint-plugin-asciidoc, textlint-plugin-asciidoc-loose). How do you think? should I  contribute to existing plugins instead of developing a new plugin?

BTW: I found some style error in your [azu/promises-book](https://www.github.com/azu/promises-book) with textlint-plugin-asciidoctor (sorry, I used the repository to verify my plugin works fine or not). Would [these commits](https://github.com/azu/promises-book/compare/master...seikichi:change-textlint-plugin?expand=1) acceptable for you? if then, I'll create a PR.

Finally, I always use textlint at work (.adoc). Thank you!

**Comment by azu** [comment]
@seikichi Great!

> should I contribute to existing plugins instead of developing a new plugin?

It is good thing that a developer who has a motivation develop plugin.
I'm ok with new plugin ❤️ 

> To verify my plugin works fine or not

https://github.com/opendevise/asciidoc-samples is good example of asciidoctor.
It will help to use snapshot testing.
Can I submit PR?

> Would these commits acceptable for you? if then, I'll create a PR.

Yes! Welcome to submit PR.

**Comment by seikichi** [comment]
Hi @azu 

Thank you for your advice and PR ( https://github.com/seikichi/textlint-plugin-asciidoctor/pull/1 ) :bow:
I did not know the https://github.com/opendevise/asciidoc-samples repository (therefore, I used your promise book and some documents in my job :-p). it will help me definitely. Thanks!

**Comment by mojavelinux** [comment]
> there are already some plugins, there are not maintained...textlint-plugin-asciidoc

I have every intention of maintaining this plugin. While it's possible to create a validator with Asciidoctor.js, it's never going to be able to go far enough. The only way to get true validation for AsciiDoc is to create a parser from scratch dedicated to validation.

I've had to deal with some other things before I could start working on it again. I estimate I should be able to start working on it again in the next month or two.

**Comment by seikichi** [comment]
@mojavelinux I'm sorry if i made you feel bad. 

> While it's possible to create a validator with Asciidoctor.js, it's never going to be able to go far enough. 

Yes, I realized that validator with Asciidoctor.js can only provide limited features. So I’m looking forward to your plugin!! (and new parser)

**Comment by mojavelinux** [comment]
No need to apologize. I just wanted to make my intentions clear so you're kept in the loop.

I actually made a bad assumption for years thinking that I was going to be able to use the main Asciidoctor parser to do validation. However, when I finally got around to trying it, I realized very quickly that it throws away too much information to really get much further than parsing blocks. That's not to say such an approach isn't valuable. Any validation is better than no validation. But I think it's time to start getting really serious about robust validation, which is why I'm suggesting a new parser. It's actually not at difficult as it seems, and it will really have a positive impact on AsciiDoc once it exists. It may inform a better parser in Asciidoctor itself. That's my hope ;)

**Comment by mojavelinux** [comment]
Of course, anything you learn in creating a parser, however you do it, that's valuable too. So I look forward to hearing your experiences.

**Comment by dgrahn** [comment]
Hi all! My apologies for resurrecting this thread.

I see that there has been little progress. Not an insult or accusation, just an acknowledgement. I know everybody has extremely full plates. Over the past few years, we've seen a lot of progress in the Asciidoc and tech space. A few things stand out to me:

1. Asciidoc has a specification an TCK that are mature.
2. Automation has come a long way.
3. Writing parsers is annoying and difficult.

And I have these questions:

1. Has Asciidoctor.js' AST evolved in a way that would make this easier?
2. Is there any appetite for someone in working on this?
3. If there was JS Asciidoc parser passed the TCK, would that make this plugin easier?
4. Would anyone be available to maintain a plugin if it was created and handed off as stable?

Thanks,
